sdk/python/kfp/compiler/_container_builder.py

```diff
@@ -0,0 +1,110 @@
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import tarfile
+import tempfile
+import os
+import uuid
+
+class ContainerBuilder(object):
+  """
+  ContainerBuilder helps build a container image
+  """
+  def __init__(self, gcs_staging):
+    if not self._check_gcs_path(gcs_staging):
+      raise Exception('ContainerBuilder __init__ failure.')
+    self._gcs_staging = gcs_staging
+
+  def _check_gcs_path(self, gcs_path):
+    """ _check_gcs_path check both the path validity and write permissions """
+    logging.info('Checking path: {}...'.format(gcs_path))
+    if not gcs_path.startswith('gs://'):
+      logging.error('Error: {} should be a GCS path.'.format(gcs_path))
+      return False
+    return True
+
+  def _generate_kaniko_spec(self, namespace, context, docker_filename, target_image):
+    """_generate_kaniko_yaml generates kaniko job yaml based on a template yaml """
+    content = {
+        'apiVersion': 'v1',
+        'metadata': {
+            'generateName': 'kaniko-',
+            'namespace': namespace,
+        },
+        'kind': 'Pod',
+        'spec': {
+            'restartPolicy': 'Never',
+            'containers': [{
+                'name': 'kaniko',
+                'args': ['--cache=true',
+                         '--dockerfile=' + docker_filename,
+                         '--context=' + context,
+                         '--destination=' + target_image],
+                'image': 'gcr.io/kaniko-project/executor@sha256:78d44ec4e9cb5545d7f85c1924695c89503ded86a59f92c7ae658afa3cff5400',
+                'env': [{
+                    'name': 'GOOGLE_APPLICATION_CREDENTIALS',
+                    'value': '/secret/gcp-credentials/user-gcp-sa.json'
+                }],
+                'volumeMounts': [{
+                    'mountPath': '/secret/gcp-credentials',
+                    'name': 'gcp-credentials',
+                }],
+            }],
+            'volumes': [{
+                'name': 'gcp-credentials',
+                'secret': {
+                    'secretName': 'user-gcp-sa',
+                },
+            }],
+            'serviceAccountName': 'default'}
+    }
+    return content
+
+  def _wrap_dir_in_tarball(self, tarball_path, dir_name):
+    """ _wrap_files_in_tarball creates a tarball for all the files in the directory"""
+    old_wd = os.getcwd()
+    os.chdir(dir_name)
+    if not tarball_path.endswith('.tar.gz'):
+      raise ValueError('the tarball path should end with .tar.gz')
+    with tarfile.open(tarball_path, 'w:gz') as tarball:
+      for f in os.listdir(dir_name):
+        tarball.add(f)
+    os.chdir(old_wd)
+
+
+  def build(self, local_dir, docker_filename, target_image, timeout, namespace):
```

Comment:
Is `docker_filename` relative to `local_dir`? I think it's a good idea to make it so.
`timeout` and `namespace` should probably be moved to the constructor.

Reply:
docker_filename is the dockerfile name that is stored in the local_dir. It should be only the filename.
I added the args comments to illustrate the input params.